managers/asset-operations-manager.js

```diff
@@ -293,23 +306,90 @@ export default class AssetOperationsManager {
             minimumNumberOfNodeReplications,
         );
 
-        let dataset;
-
+        let dataset = {};
         if (typeof content === 'string') {
-            dataset = content
-                .split('\n')
-                .map((line) => line.trimStart().trimEnd())
-                .filter((line) => line.trim() !== '');
+            dataset.public = this.processContent(content);
+        } else if (
+            typeof content.public === 'string' ||
+            (!content.public && content.private && typeof content.private === 'string')
+        ) {
+            if (content.public) {
+                dataset.public = this.processContent(content.public);
+            }
+            if (content.private && typeof content.private === 'string') {
+                dataset.private = this.processContent(content.private);
+            }
         } else {
             dataset = await kcTools.formatDataset(content);
         }
 
-        const numberOfChunks = kcTools.calculateNumberOfChunks(dataset, CHUNK_BYTE_SIZE);
+        dataset.public = kcTools.generateMissingIdsForBlankNodes(dataset.public);
+        if (dataset.private) {
+            dataset.private = kcTools.generateMissingIdsForBlankNodes(dataset.private);
+        }
+
+        let privateTripletsSortedAndGrouped;
+        const publicTripletsSortedAndGrouped = kcTools.groupNquadsBySubject(dataset.public, true);
+        if (dataset.private) {
+            privateTripletsSortedAndGrouped = kcTools.groupNquadsBySubject(dataset.private, true);
+        }
+
+        let publicIndex = 0;
+        let privateIndex = 0;
+
+        while (
+            publicIndex < publicTripletsSortedAndGrouped.length ||
+            privateIndex < privateTripletsSortedAndGrouped.length
+        ) {
+            if (publicIndex === publicTripletsSortedAndGrouped.length) {
+                const [privateSubject] =
+                    privateTripletsSortedAndGrouped[privateIndex][0].split(' ');
+                const privateMerkleRootTriplet = `${privateSubject} <${PRIVATE_ASSERTION_PREDICATE}> "${kcTools.calculateMerkleRoot(
+                    privateTripletsSortedAndGrouped[privateIndex],
+                )}" .`;
+                publicTripletsSortedAndGrouped.push([privateMerkleRootTriplet]);
+                privateIndex += 1;
+                continue;
+            }
+            if (privateIndex === privateTripletsSortedAndGrouped.length) {
+                // Only public left no need to do anything
+                break;
+            }
+            const [publicSubject] = publicTripletsSortedAndGrouped[publicIndex][0].split(' ');
+            const [privateSubject] = privateTripletsSortedAndGrouped[privateIndex][0].split(' ');
+            let compare = publicSubject.localeCompare(privateSubject);
+            if (compare === -1) {
+                // Public comes next
+                publicIndex += 1;
```

Comment:
you have public index in all parts of if statement, you can put it outside of it probably

Reply:
That's correct, but I think this is easier to read so you have logic for each case there in that block